pasha/functor.py

```diff
@@ -0,0 +1,280 @@
+
+# Distributed under the terms of the BSD 3-Clause License.
+# The full license is in the file LICENSE, distributed with this software.
+#
+# Author: Philipp Schmidt <philipp.schmidt@xfel.eu>
+# Copyright (c) 2020, European X-Ray Free-Electron Laser Facility GmbH.
+# All rights reserved.
+
+import sys
+from collections.abc import Sequence
+
+import numpy as np
+
+
+def gen_split_slices(total_len, part_len=None, n_parts=None):
+    """Generate slices to split a sequence.
+
+    Returns a list of slices to split a sequence into smaller pieces.
+
+    Args:
+        total_len (int): Length of the full sequence.
+        part_len (int, optional): Length of each piece, may be None if
+            n_parts is specified.
+        n_parts (int, optional): Number of pieces, ignored if part_len
+            is specified.
+
+    Returns:
+        (list of slice) Slices to split the sequence.
+    """
+
+    if part_len is None:
+        if n_parts is None:
+            raise ValueError('must specify either part_len or n_parts')
+
+        part_len = total_len // n_parts
+
+    return [slice(start, min(total_len, start + part_len), 1)
+            for start in range(0, total_len, part_len)]
```

Comment:
It looks like you only use this with `n_parts`, but FWIW, it may make an unbalanced split if used with `part_len`. Here's some code I've written previously to do a balanced split by treating the chunk size as a maximum:

```python
n_chunks = math.ceil(N / chunk_size)
for i in range(n_chunks):
    yield i * N // n_chunks, (i+1) * N // n_chunks
```

Don't know if that's important, it just came to mind when I saw this. :-)

Reply:
My code probably treats `part_len` to be holier than necessary. The code you pasted produces nicer results in a few edge cases, I want it. You've got some project I can quote it from?

Reply:
It's in pycalibration, but I wrote it ([in this commit](https://in.xfel.eu/gitlab/detectors/pycalibration/merge_requests/330/diffs?commit_id=147d964ce10cf442e88a42156c0b9c7e028ee05a)), and I'm happy to contribute it to pasha as well.

Reply:
Made a PR (#2) for it. :slightly_smiling_face: